http_proxy.go

```diff
@@ -960,6 +966,19 @@ func (p *Proxy) listenAlgeneva(baseListen func(string) (net.Listener, error)) li
 	}
 }
 
+func (p *Proxy) listenWATERReverse() listenerBuilderFN {
+	return func(addr string) (net.Listener, error) {
+		ctx := context.Background()
+		reverseListener, err := water.NewReverseListener(ctx, addr, p.WaterWASM)
+		if err != nil {
+			return nil, err
+		}
+
+		log.Debugf("Listening for water at %v", reverseListener.Addr())
+		return reverseListener, nil
+	}
+}
```

Comment:
Could you wrap a base listener like the other transports do?  [Here's an example](https://github.com/getlantern/http-proxy/blob/0df4a97b806e9a9078b9ae07da21c2bd04ff1b8a/http_proxy.go#L933).  It looks like you should be able to use [water.Config.Listener](https://pkg.go.dev/github.com/refraction-networking/water#Config) to achieve this, unless I'm missing something.

Reply:
The `water.Config.NetworkListener` attribute is overrided when we call  the function for starting listening with [`cfg.ListenContext`](https://github.com/refraction-networking/water/blob/v0.6.4/config.go#L137-L151) 😢 (as mentioned [here](https://pkg.go.dev/github.com/refraction-networking/water#Config)):

> // NetworkListener specifies a net.listener implementation that listens
	// on the specified address on the named network. This optional field
	// will be used to provide (incoming) network connections from a
	// presumably remote source to the WASM instance.
	//
	// Calling (*Config).Listen will override this field.
	NetworkListener [net](https://pkg.go.dev/net).[Listener](https://pkg.go.dev/net#Listener)

That's why I didn't wrap with the base listener (I'll need another suggestion)

Reply:
By providing a listener, without calling the `ListenContext` we would be basically using the listener that doesn't support or accept a connection with the desired protocol (from WASM)

Reply:
Oh sorry, I missed that you were using `Config.ListenContext`.  It seems like [water.NewListenerWithContext](https://pkg.go.dev/github.com/refraction-networking/water#NewListenerWithContext) is supposed to use the configured listener.  However, looking through the code and running some tests myself, that does not appear to be the case.  It looks like `Config.NetworkListener` is actually never called anywhere in the `water` code.  I'm assuming that's an oversight - I'll reach out to the WATER team about it.

> By providing a listener, without calling the ListenContext we would be basically using the listener that doesn't support or accept a connection with the desired protocol (from WASM)

Just for clarity: the idea is to provide a base transport.  We would still use the protocol configured in the WASM.  For example, when we create our algeneva listener, we specify that it run over a custom TCP listener.  In theory we could run algeneva over QUIC, TLS, or any other reliable transport.

The reason we want to use a custom TCP listener is just to [control things like idle timeouts](https://github.com/getlantern/http-proxy/blob/0df4a97b806e9a9078b9ae07da21c2bd04ff1b8a/http_proxy.go#L758-L768).

Reply:
>  I'm assuming that's an oversight - I'll reach out to the WATER team about it.

To be clear, I think we should go with what you've got here for now.  If the WATER team fixes the network listener issue, we can update the code.  I imagine it might be tricky for them though (I was pleasantly surprised to see that the API appeared to support a custom base listener to begin with).

Reply:
aaaa looking into NewListenerWithContext (without being from the config) there's some examples here for wrapping a listener: https://github.com/refraction-networking/water/blob/master/transport/v1/listener_test.go#L18-L34

I'll add that

Reply:
Awesome, thanks for doing that!  Also looks like I was wrong on there being a bug in wrapping listeners.  I found an issue in my test and was able to get it to work.